src/self_driving/forecasting/workload_forecast.cpp

```diff
@@ -14,20 +14,50 @@
 
 namespace noisepage::selfdriving {
 
-WorkloadForecast::WorkloadForecast(uint64_t forecast_interval) : forecast_interval_(forecast_interval) {
+WorkloadForecast::WorkloadForecast(uint64_t forecast_interval, uint64_t num_sample)
+    : num_sample_(num_sample), workload_metadata_({}), forecast_interval_(forecast_interval) {
   LoadQueryText();
   LoadQueryTrace();
   CreateSegments();
 }
 
+void WorkloadForecast::InitFromInference(const WorkloadForecastPrediction &inference) {
+  bool init = false;
+  for (auto &cluster : inference) {
+    for (auto &query_info : cluster.second) {
+      if (!init) {
+        forecast_segments_ = std::vector<WorkloadForecastSegment>(query_info.second.size());
+        init = true;
+      }
+
+      for (size_t seg = 0; seg < query_info.second.size(); seg++) {
+        double freq = query_info.second[seg];
+        forecast_segments_[seg].AddQueryInfo(execution::query_id_t{static_cast<uint32_t>(query_info.first)}, freq);
+      }
+    }
+  }
+  num_forecast_segment_ = forecast_segments_.size();
+}
+
+WorkloadForecast::WorkloadForecast(const WorkloadForecastPrediction &inference, WorkloadMetadata &&metadata) {
+  workload_metadata_ = std::move(metadata);
+  InitFromInference(inference);
+}
+
+WorkloadForecast::WorkloadForecast(const WorkloadForecastPrediction &inference) {
```

Comment:
I figured out why we're executing much more queries in this mode. Essentially we forgot to initialize `num_sample_` in this constructor, and we're loading all the parameters from `query_trace.csv`. We should hook this up with the `forecast_sample_limit` knob.

Reply:
Ah. Thanks for the catch!